rasa/core/events/__init__.py

```diff
@@ -632,87 +651,143 @@ def __eq__(self, other) -> bool:
             )
         ]
 
 
 # noinspection PyProtectedMember
 class ReminderCancelled(Event):
-    """Cancel all jobs with a specific name."""
+    """Cancel certain jobs."""
 
     type_name = "cancel_reminder"
 
     def __init__(
         self,
-        action_name: Text,
+        name: Optional[Text] = None,
+        intent: Optional[Text] = None,
+        entities: Optional[List[Dict]] = None,
         timestamp: Optional[float] = None,
         metadata: Optional[Dict[Text, Any]] = None,
     ):
-        """
+        """Creates a ReminderCancelled event.
+
+        All arguments are optional and used to refine the set of reminders that
+        are to be cancelled. If no arguments are supplied, this will cancel all reminders.
+
         Args:
-            action_name: name of the scheduled action to be cancelled
-            metadata: optional event metadata
+            name: Name of the reminder to be cancelled.
+            intent: Intent name that is to be used to identify the reminders to be cancelled.
+            entities: Entities that are to be used to identify the reminders to be cancelled.
+            timestamp: Optional timestamp.
+            metadata: Optional event metadata.
         """
 
-        self.action_name = action_name
+        self.name = name
+        self.intent = intent
+        self.entities = entities
         super().__init__(timestamp, metadata)
 
     def __hash__(self) -> int:
-        return hash(self.action_name)
+        return hash((self.name, self.intent, str(self.entities),))
 
     def __eq__(self, other) -> bool:
-        return isinstance(other, ReminderCancelled)
+        if not isinstance(other, ReminderCancelled):
+            return False
+        else:
+            return hash(self) == hash(other)
 
     def __str__(self) -> Text:
-        return f"ReminderCancelled(action: {self.action_name})"
+        return f"ReminderCancelled(name: {self.name}, intent: {self.intent}, entities: {self.entities})"
+
+    def cancels_job_with_name(self, job_name: Text, sender_id: Text) -> bool:
+        """Determines if this ReminderCancelled should cancel the job with the given name.
+
+        Args:
+            job_name: Name of the job to be tested.
+            sender_id: The `sender_id` of the tracker.
+
+        Returns:
+            `True`, if this `ReminderCancelled` should cancel the job with the given name,
+            and `False` otherwise.
+        """
+
+        match = re.match(
+            rf"^\[([\d\-]*),([\d\-]*),([\d\-]*)\]"
+            + rf"({re.escape(ACTION_NAME_SENDER_ID_CONNECTOR_STR)}{re.escape(sender_id)})",
+            job_name,
+        )
+        if not match:
+            return False
+        name_hash, intent_hash, entities_hash = match.group(1, 2, 3)
+
+        # Cancel everything unless names/intents/entities are given to
+        # narrow it down.
+        return (
+            ((not self.name) or str(hash(self.name)) == name_hash)
```

Comment:
optional: use helper functions to make the condition easier understandable
```suggestion
            is_name_equal(self.name, name_hash) and is_intent_equal(self.intent, intent_hash) and are_entities_equal(self.entities, entities_hash)
```